handler/oauth2/flow_generic_code_token.go

```diff
@@ -0,0 +1,244 @@
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+	defer func() {
+		if err != nil {
+			if rollBackTxnErr := storage.MaybeRollbackTx(ctx, c.CoreStorage); rollBackTxnErr != nil {
+				err = errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebugf("error: %s; rollback error: %s", err, rollBackTxnErr))
+			}
+		}
+	}()
+
+	if err = c.InvalidateSession(ctx, signature); err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	if err = c.CoreStorage.CreateAccessTokenSession(ctx, accessTokenSignature, requester.Sanitize([]string{})); err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	if refreshTokenSignature != "" {
+		if err = c.CoreStorage.CreateRefreshTokenSession(ctx, refreshTokenSignature, requester.Sanitize([]string{})); err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
+	}
+
+	lifeSpan := fosite.GetEffectiveLifespan(requester.GetClient(), fosite.GrantTypeAuthorizationCode, fosite.AccessToken, c.Config.GetAccessTokenLifespan(ctx))
+	responder.SetAccessToken(accessToken)
+	responder.SetTokenType("bearer")
+	responder.SetExpiresIn(getExpiresIn(requester, fosite.AccessToken, lifeSpan, time.Now().UTC()))
+	responder.SetScopes(requester.GetGrantedScopes())
+	if refreshToken != "" {
+		responder.SetExtra("refresh_token", refreshToken)
+	}
+
+	if err = storage.MaybeCommitTx(ctx, c.CoreStorage); err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return nil
+}
+
+func (c *GenericCodeTokenEndpointHandler) HandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(ctx, requester) {
+		return errorsx.WithStack(errorsx.WithStack(fosite.ErrUnknownRequest))
+	}
+
+	if err := c.ValidateGrantTypes(ctx, requester); err != nil {
+		return err
+	}
+
+	code, _, ar, err := c.GetCodeAndSession(ctx, requester)
+	if err != nil {
+		switch {
+		case errors.Is(err, fosite.ErrInvalidatedAuthorizeCode), errors.Is(err, fosite.ErrInvalidatedDeviceCode):
+			if ar == nil {
+				return fosite.ErrServerError.
+					WithHint("Misconfigured code lead to an error that prohibited the OAuth 2.0 Framework from processing this request.").
+					WithDebug("getCodeSession must return a value for \"fosite.Requester\" when returning \"ErrInvalidatedAuthorizeCode\" or \"ErrInvalidatedDeviceCode\".")
+			}
+
+			return c.revokeTokens(ctx, requester.GetID())
+		case errors.Is(err, fosite.ErrAuthorizationPending):
+			return err
+		case errors.Is(err, fosite.ErrPollingRateLimited):
+			return errorsx.WithStack(err)
+		case errors.Is(err, fosite.ErrNotFound):
+			return errorsx.WithStack(fosite.ErrInvalidGrant.WithWrap(err).WithDebug(err.Error()))
+		default:
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
```

Comment:
Can we add a method in `CodeTokenEndpointHandler` to handle errors (e.g. `HandleError)